Add RBAC good practice guide (#32812)
kubernetes/website: content/en/docs/concepts/security/rbac-good-practices.md

---
reviewers:

> title: RBAC Good Practices

Reviewer: I like having it in Concepts. To me that's a closer fit than Reference: reference information should always be objective, whereas good practice is often subjective.

> title: RBAC Good Practices

Author: I'm fine with either but yep I can see this being more concepts

description: >

>   Principles and practices for good RBAC design for cluster operators.

Reviewer: In this title, consider expanding the abbreviation “RBAC”.

Author: expanded

Reviewer: That works (I was actually thinking of expanding “Principles and practices for good RBAC design for cluster operators.”).

content_type: concept
weight: 10
---

> # Kubernetes RBAC Good Practices

Reviewer: Please omit this; the Hugo theme that we use renders a first-level heading based on the title you set in the front matter.

Author: removed

<!-- overview -->

> Kubernetes RBAC is a key security control to ensure that cluster users and workloads have only the access to resources required to execute their roles. It is important to ensure that, when designing permissions for cluster users, the cluster administrator understands the areas where privilge escalation could occur, to reduce the risk of excessive access leading to security incidents.

Reviewer: I would use a [glossary tooltip shortcode](https://kubernetes.io/docs/contribute/style/hugo-shortcodes/#glossary) for RBAC.

Author: Added , think I got it right :)

> Kubernetes RBAC is a key security control to ensure that cluster users and workloads have only the access to resources required to execute their roles. It is important to ensure that, when designing permissions for cluster users, the cluster administrator understands the areas where privilge escalation could occur, to reduce the risk of excessive access leading to security incidents.

Author: righto, is there a specific length I should shoot for (e.g. 80 chars)?

The good practices laid out here should be read in conjunction with the general [RBAC documentation](/docs/reference/access-authn-authz/rbac/#restrictions-on-role-creation-or-update).

<!-- body -->

> ## General Good Practice

Reviewer: nit: page headings should be in sentence case

Author: Sentence cased the titles

### Least Privilege

Ideally minimal RBAC rights should be assigned to users and service accounts. Only permissions explicitly required for their operation should be used. Whilst each cluster will be different, some general rules that can be applied are :-

>  - Assign permissions at the namespace level where possible. Use rolebindings as opposed to clusterrolebindings to give users rights only within a specific namespace.

Author: updated

### Minimize Distribution of Privileged Tokens

Ideally, pods shouldn't be assigned service accounts granted powerful permissions (listed [here](#Kubernetes-RBAC---Privilege-Escalation-Risks)). In cases where a workload requires powerful permissions, consider the following best practices:

 - Limit the number of nodes running powerful pods. Avoid running powerful Daemonsets to ensure nodes are least-privileged and to limit the blast radius of container escapes.
 - Avoid running powerful pods alongside untrusted or publicly-exposed ones. Consider using [Taints and Toleration](/docs/concepts/scheduling-eviction/taint-and-toleration/), [NodeAffinity](/docs/concepts/scheduling-eviction/assign-pod-node/#node-affinity), or [PodAntiAffinity](/docs/concepts/scheduling-eviction/assign-pod-node/#inter-pod-affinity-and-anti-affinity) to ensure powerful pods don't run alongside untrusted ones.

### Hardening

> Kubernetes defaults to providing access which may not be required in every cluster. Reviewing the RBAC rights provided by default can provide opportunities for security hardening. Whilst, in general, changes should not be made to rights provided to `system:` accounts some options to harden cluster rights exist :- 

Reviewer: I would use a plain colon (`:`) to introduce the list.

- Review bindings for the `system:unauthenticated` group and remove where possible, as this gives access to anyone who can contact the API server at a network level.

> - Set service accounts not to mount tokens by default using the `automountServiceAccountToken: false` setting (additional information [here]/docs/tasks/configure-pod-container/configure-service-account/#use-the-default-service-account-to-access-the-api-server)). As setting this value in a pod will overwrite the service account setting, workloads which require service account tokens can still mount them.

Reviewer 2: ```suggestion
- Avoid the default auto-mounting of service account tokens by setting
  `automountServiceAccountToken: false`. For more details, see
  [using default service account token]/docs/tasks/configure-pod-container/configure-service-account/#use-the-default-service-account-to-access-the-api-server).
  Setting this value for a Pod will overwrite the service account setting, workloads
  which require service account tokens can still mount them.
```

### Periodic Review

> It is vital to periodically review Kubernetes RBAC assignments for redundant entries, to avoid possible privilege escalation issues. If an attacker is able to create a user account with the same name as a deleted one, they will automatically inherit all the rights that that previous account had (where the right was assigned to the username specifically).

Reviewer 2: ```suggestion
It is vital to periodically review the Kubernetes RBAC settings for redundant entries and possible privilege escalations.
If an attacker is able to create a user account with the same name as a deleted user,
he or she can automatically inherit all the rights of the deleted user, especially the
rights assigned to that user.
```

Reviewer: If we can, avoid implying that “he” and “she” form an exhaustive list of personal pronouns. We have end users and indeed contributors who prefer different pronouns.

## Kubernetes RBAC - Privilege Escalation Risks 

> Within Kubernetes RBAC there are a number of privileges which, if granted, can allow for a user or service account to escalate their rights in the cluster or affect systems outside the cluster.

Reviewer 2: ```suggestion
Within Kubernetes RBAC there are a number of privileges which, if granted, can allow a user or a service account
to escalate their privileges in the cluster or affect systems outside the cluster.
```

Reviewer: Consider whether it is helpful to write ServiceAccount here (PascalCase) - that can help readers understand that user is **not** a Kubernetes API, whereas ServiceAccount is.

This document is intended to provide visibility of the areas where cluster administrators should take care, to ensure that they do not inadvertantly allow for more access to clusters than intended.

> ### Secrets Objects - List

Reviewer 2: ```suggestion
### Listing Secrets
```

> Whilst it is generally clear that allowing **GET** access to secrets will allow a user to read their contents, which then allows for them to be re-used, it's important to note that **LIST** access effectively also allows for this, as when a list request is returned (e.g. via `kubectl get secrets -A -o yaml`) it will include the contents of the secret.

Reviewer 2: ```suggestion
It is generally clear that allowing **Get** access on Secrets will allow a user to read their contents.
It is also important to note that the **List** access also effectively allows for users to reveal the Secret contents.
When a List response is returned (for example, via `kubectl get secrets -A -o yaml`), the response
includes the contents of all Secrets.
```

Reviewer: If it's an HTTP GET, write `GET` in code style; if it's **get** at the Kubernetes API level, write it lower case and bold. A similar story with **list**.

Aside: the response to a **list** request is usually a _collection_. If you list Secrets, you get back a collection of Secrets.

### Workload Creation

> Users who are able to create workloads (e.g. pods) will be able to gain access to the underlying node if restrictions based on the [Pod Security Standards](/docs/concepts/security/pod-security-standards/) are not in place. Baseline or Restricted profiles should be used for all users with rights to create workloads in the cluster.

Reviewer: ```suggestion
Users who are able to create workloads (either Pods, or
[workload resources](/docs/concepts/workloads/controllers/) that manage Pods) will
be able to gain access to the underlying node unless restrictions based on the Kubernetes
[Pod Security Standards](/docs/concepts/security/pod-security-standards/) are in place.

Users who can run privileged Pods can use that access to gain node access and potentially to
further elevate their privileges. Where you do not fully trust a user or other principal
with the ability to create suitably secure and isolated Pods, you should enforce either the
**Baseline** or **Restricted** Pod Security Standard.
You can use [Pod Security admission](/docs/concepts/security/pod-security-admission/)
or other (third party) mechanisms to implement that enforcement.

You can also use the deprecated [PodSecurityPolicy](/docs/concepts/policy/pod-security-policy/) mechanism
to restrict users' abilities to create privileged Pods.
```

> Creating a workload in a namespace also grants indirect access to secrets in that namespace. Creating a pod in kube-system or a similarly privileged namespace can grant a user access to secrets they would not have through RBAC directly.  

Reviewer: ```suggestion
Creating a workload in a namespace also grants indirect access to Secrets in that namespace. Creating a pod in kube-system or a similarly privileged namespace can grant a user access to Secrets they would not have through RBAC directly.
```

> ### Persistent Volume Creation

Reviewer: ```suggestion
### PersistentVolume creation
```

> As noted in the [Pod Security Policy](/docs/concepts/policy/pod-security-policy/#volumes-and-file-systems) documentation, access to create Persistent Volumes can allow for escalation of access to the underlying host.

Reviewer: ```suggestion
As noted in the [PodSecurityPolicy](/docs/concepts/policy/pod-security-policy/#volumes-and-file-systems) documentation, access to create PersistentVolumes can allow for escalation of access to the underlying host.
```

### Node/Proxy Access

> Users with access to the proxy sub-resource of node objects have rights to the Kubelet API, which allows for command execution on every pod on the node(s) which they have rights to. This access bypasses audit logging and admission control, so care should be taken before granting rights to this resource.

Reviewer: ```suggestion
Users with access to the proxy sub-resource of node objects have rights to access the kubelet's HTTP API, proxying through your cluster's API server. That proxied access allows for command execution on any and every pod on the node(s) to which a user has access.

{{< caution >}}
Proxied requests to the kubelet bypass Kubernetes audit logging and admission control, so you
should take special care before granting access to this subresource.
{{< /caution >}}
```

### Escalate Right

> Generally the RBAC system prevents users from creating clusterroles with more rights than they possess. The exception to this is the escalate right. As noted in the [RBAC documentation](/docs/reference/access-authn-authz/rbac/#restrictions-on-role-creation-or-update) users with this right can effectively escalate their privileges.

Reviewer: ```suggestion
The RBAC system prevents users from creating ClusterRoles with more rights than they possess.
There is an explicit exception to that preventative mechanism: the `escalate` right. As noted in the [RBAC documentation](/docs/reference/access-authn-authz/rbac/#restrictions-on-role-creation-or-update) users with
this right can effectively escalate their privileges, because they are authorized to do so.
```

I would overall try to frame it that the `escalate` right is equivalent _within RBAC_ to full access.
We might want to hint that other mechanisms&mdash;a webhook authorizer, for example&mdash;can further
control a principal's ability to elevate their privileges.

### Bind Right

Similar to the Escalate right, granting users this right allows for bypass of Kubernetes in-built protections against privilege escalation, allowing users to create bindings to roles with rights they do not already have.

### Impersonate Right

This right allows users to impersonate and gain the rights of other users in the cluster. Care should be taken when granting it, to ensure that excessive permissions cannot be gained via one of the impersonated accounts.

### CSRs and certificate issuing

The CSR API allows for users with `create` rights to CSRs and `update` rights on `certificatesigningrequests/approval` where the signer is `kubernetes.io/kube-apiserver-client` to create new client certificates which allow users to authenticate to the cluster. Those client certificates can have arbitrary names including duplicates of Kubernetes system components. This will effectively allow for privilege escalation.

> ### Token Request

Reviewer: Maybe:
```suggestion
### TokenRequest subresource for Secret
```

Users with `create` rights on `serviceaccounts/token` can create TokenRequests to issue tokens for existing service accounts. 

> ### Control Admission Webhooks

Reviewer: ```suggestion
### Restrictions on admission webhooks
```

> Users with control over `validatingwebhookconfigurations` or `mutatingwebhookconfigurations` can control webhooks that can read any object admitted to the cluster, and in the case of mutating webhooks, also mutate admitted objects.

Reviewer: - I'd write ValidatingWebhookConfiguration and MutatingWebhookConfiguration in the text; however, you could also give an example of a dangerous RoleBinding (use a caution shortcode and inline comments) than mentions the API path fragment, eg `mutatingwebhookconfigurations`).
- if possible, [address the reader as you](https://kubernetes.io/docs/contribute/style/style-guide/#address-the-reader-as-you)

> ## Kubernetes RBAC - Denial of Service Risks

Reviewer: ```suggestion
## Denial of Service risks that relate to RBAC
```

### Object Creation DoS

> Users who have rights to create objects in a cluster may be able to create sufficient large objects to create a denial of service condition, as discussed [here](https://github.com/kubernetes/kubernetes/issues/107325). This may be specifically relevant in multi-tenant clusters if semi-trusted or untrusted users are allowed limited access to a system.

Reviewer: ```suggestion
If you have rights to create objects in a cluster, you may be able to create sufficient large objects
to create a denial of service condition, as discussed in Kubernetes
[issue 107325](https://github.com/kubernetes/kubernetes/issues/107325). This risk may be specifically
relevant in multi-tenant clusters if semi-trusted or untrusted users are allowed limited access to a system;
someone with only partial access may be able to use that access to deny service to the cluster overall.
```

 - Avoid providing wildcard permissions when possible, especially to all resources.
   As Kubernetes is an extensible system, providing wildcard access gives rights
   not just to all object types presently in the cluster, but also to all future object types
   which are created in the future.
 - Administrators should not use `cluster-admin` accounts except where specifically needed. 
   Providing a low privileged account with impersonation rights can avoid accidental modification of 
   cluster resources.
 - Avoid adding users to the `system:masters` group. Any user who is a member of this group 
   bypasses all RBAC rights checks and will always have cluster-admin access, which cannot be 
   revoked by removing Role Bindings or Cluster Role Bindings. As an aside, if a cluster is 
   using an authorization webhook, membership of this group also bypasses that webhook (requests 
   from users who are members of that group are never sent to the webhook)

### Minimize distribution of privileged tokens

Ideally, pods shouldn't be assigned service accounts granted powerful permissions (listed [here](#Kubernetes-RBAC---Privilege-Escalation-Risks)). 

> In cases where a workload requires powerful permissions, consider the following best practices:

Author: agreed yeah I generally avoid best, must've missed this one.

the RBAC rights provided by default can provide opportunities for security hardening. Whilst, 
in general, changes should not be made to rights provided to `system:` accounts some options 
to harden cluster rights exist :- 

- Review bindings for the `system:unauthenticated` group and remove where possible, as this gives 
  access to anyone who can contact the API server at a network level.
- Avoid the default auto-mounting of service account tokens by setting
  `automountServiceAccountToken: false`. For more details, see
  [using default service account token]/docs/tasks/configure-pod-container/configure-service-account/#use-the-default-service-account-to-access-the-api-server).
  Setting this value for a Pod will overwrite the service account setting, workloads
  which require service account tokens can still mount them.

### Periodic review

It is vital to periodically review the Kubernetes RBAC settings for redundant entries and 
possible privilege escalations.
If an attacker is able to create a user account with the same name as a deleted user,
they can automatically inherit all the rights of the deleted user, especially the
rights assigned to that user.

> ## Kubernetes RBAC - privilege escalation risks 

Reviewer: or:
```suggestion
## Kubernetes RBAC - privilege escalation risks {#privilege-escalation-risks}
```

---
reviewers:
title: Role Based Access Control Good Practices
description: >
  Principles and practices for good RBAC design for cluster operators.
content_type: concept

> weight: 10

Reviewer: nit: Omit this.

The important outcome is that this page should sort _after_ “Controlling Access To The Kubernetes API” in the navigation.

Author: can I just remove it altogether, or just change the number?

Reviewer: Remove it all together

Author: cool, done :)

to reduce the risk of excessive access leading to security incidents.

The good practices laid out here should be read in conjunction with the general [RBAC documentation](/docs/reference/access-authn-authz/rbac/#restrictions-on-role-creation-or-update).

<!-- body -->

## General good practice

### Least privilege

Ideally minimal RBAC rights should be assigned to users and service accounts. Only permissions 
explicitly required for their operation should be used. Whilst each cluster will be different, 
some general rules that can be applied are :

 - Assign permissions at the namespace level where possible. Use RoleBindings as opposed to 
   ClusterRoleBindings to give users rights only within a specific namespace.
 - Avoid providing wildcard permissions when possible, especially to all resources.
   As Kubernetes is an extensible system, providing wildcard access gives rights
   not just to all object types presently in the cluster, but also to all future object types
   which are created in the future.
 - Administrators should not use `cluster-admin` accounts except where specifically needed. 

>    Providing a low privileged account with impersonation rights can avoid accidental modification of 

Author: Cool, added :)

   cluster resources.
 - Avoid adding users to the `system:masters` group. Any user who is a member of this group 

>    bypasses all RBAC rights checks and will always have cluster-admin access, which cannot be 

Author: Yeah good point, I went for "unrestricted superuser" which I think makes the point while avoiding the possible naming confusion.

   revoked by removing Role Bindings or Cluster Role Bindings. As an aside, if a cluster is 
   using an authorization webhook, membership of this group also bypasses that webhook (requests 
   from users who are members of that group are never sent to the webhook)

### Minimize distribution of privileged tokens

Ideally, pods shouldn't be assigned service accounts granted powerful permissions (listed [here](#Kubernetes-RBAC---Privilege-Escalation-Risks)). 
In cases where a workload requires powerful permissions, consider the following practices:

 - Limit the number of nodes running powerful pods. Ensure that any DaemonSets you run
  are necessary and are run with least privilege.

>    nodes are least-privileged and to limit the blast radius of container escapes.

Reviewer 3: copy/paste or sentence fragment?

Author: yeah  I think that one got a bit messed up somewhere. The sentence should be `Ensure that any DaemonSets you run
  are necessary and are run with least privilege to limit the blast radius of container escapes` . Updated.

 - Avoid running powerful pods alongside untrusted or publicly-exposed ones. Consider using 
   [Taints and Toleration](/docs/concepts/scheduling-eviction/taint-and-toleration/), [NodeAffinity](/docs/concepts/scheduling-eviction/assign-pod-node/#node-affinity), or [PodAntiAffinity](/docs/concepts/scheduling-eviction/assign-pod-node/#inter-pod-affinity-and-anti-affinity) to ensure 
   pods don't run alongside untrusted or less-trusted Pods. Pay especial attention to
   situations where less-trustworthy Pods are not meeting the **Restricted** Pod Security Standard.

### Hardening

Kubernetes defaults to providing access which may not be required in every cluster. Reviewing 

> the RBAC rights provided by default can provide opportunities for security hardening. Whilst, 
> in general, changes should not be made to rights provided to `system:` accounts some options 

Reviewer 3: ```suggestion
the RBAC rights provided by default can provide opportunities for security hardening.
In general, changes should not be made to rights provided to `system:` accounts some options 
```

  Setting this value for a Pod will overwrite the service account setting, workloads
  which require service account tokens can still mount them.

### Periodic review

It is vital to periodically review the Kubernetes RBAC settings for redundant entries and 
possible privilege escalations.
If an attacker is able to create a user account with the same name as a deleted user,
they can automatically inherit all the rights of the deleted user, especially the
rights assigned to that user.

## Kubernetes RBAC - privilege escalation risks 

Within Kubernetes RBAC there are a number of privileges which, if granted, can allow a user or a service account
to escalate their privileges in the cluster or affect systems outside the cluster.

This section is intended to provide visibility of the areas where cluster operators 
should take care, to ensure that they do not inadvertantly allow for more access to clusters than intended.

### Listing secrets

It is generally clear that allowing **Get** access on Secrets will allow a user to read their contents.

> It is also important to note that the **List** access also effectively allows for users to reveal the Secret contents.

Reviewer 3: and watch

Author: oh yeah that makes sense. Adapted this sentence to mention watch as well :)

When a List response is returned (for example, via `kubectl get secrets -A -o yaml`), the response
includes the contents of all Secrets.

### Workload creation

Users who are able to create workloads (either Pods, or
[workload resources](/docs/concepts/workloads/controllers/) that manage Pods) will
be able to gain access to the underlying node unless restrictions based on the Kubernetes
[Pod Security Standards](/docs/concepts/security/pod-security-standards/) are in place.

Users who can run privileged Pods can use that access to gain node access and potentially to
further elevate their privileges. Where you do not fully trust a user or other principal
with the ability to create suitably secure and isolated Pods, you should enforce either the
**Baseline** or **Restricted** Pod Security Standard.
You can use [Pod Security admission](/docs/concepts/security/pod-security-admission/)
or other (third party) mechanisms to implement that enforcement.

You can also use the deprecated [PodSecurityPolicy](/docs/concepts/policy/pod-security-policy/) mechanism

> to restrict users' abilities to create privileged Pods.

Reviewer 3: I'd avoid mentioning this in new documentation, or if we mention it, explicitly call out the 1.25 removal

Author: @sftim What's your preference here, should we just remove this bit, or I can add an explicit mention of the removal in 1.25. On the one hand removing it makes sense as it is deprecated. On the other hand I know how many really old clusters are still running for which PSP is likely a good way to address this kind of issue...

Reviewer: I would add it here and then clean the new doc up for the _dev-1.25_ branch.

Author: righto. Added a specific note of version deprecation here and we can clean in 1.25 docs :)

Creating a workload in a namespace also grants indirect access to Secrets in that namespace. 
Creating a pod in kube-system or a similarly privileged namespace can grant a user access to 
Secrets they would not have through RBAC directly.

### Persistent volume creation

> As noted in the [PodSecurityPolicy](/docs/concepts/policy/pod-security-policy/#volumes-and-file-systems) documentation, access to create PersistentVolumes can allow for escalation of access to the underlying host.

Reviewer 3: might be worth a sentence here indicating that constrained users are expected to create PersistentVolumeClaims as a way of obtaining storage safely, not PersistentVolumes directly.

Author: cool make sense, added

### Access to `proxy` subresource of Nodes

Users with access to the proxy sub-resource of node objects have rights to the Kubelet API, 
which allows for command execution on every pod on the node(s) which they have rights to. 
This access bypasses audit logging and admission control, so care should be taken before 
granting rights to this resource.

### Escalate right

Generally the RBAC system prevents users from creating clusterroles with more rights than 

> they possess. The exception to this is the `escalate` right. As noted in the [RBAC documentation](/docs/reference/access-authn-authz/rbac/#restrictions-on-role-creation-or-update),

Author: Cool, changed.

roles with rights they do not already have.

### Impersonate right

This right allows users to impersonate and gain the rights of other users in the cluster. 
Care should be taken when granting it, to ensure that excessive permissions cannot be gained 
via one of the impersonated accounts.

### CSRs and certificate issuing

The CSR API allows for users with `create` rights to CSRs and `update` rights on `certificatesigningrequests/approval` 
where the signer is `kubernetes.io/kube-apiserver-client` to create new client certificates 
which allow users to authenticate to the cluster. Those client certificates can have arbitrary 
names including duplicates of Kubernetes system components. This will effectively allow for privilege escalation.

### Token request

Users with `create` rights on `serviceaccounts/token` can create TokenRequests to issue 
tokens for existing service accounts. 

### Control admission webhooks

Users with control over `validatingwebhookconfigurations` or `mutatingwebhookconfigurations` 
can control webhooks that can read any object admitted to the cluster, and in the case of 
mutating webhooks, also mutate admitted objects.

> ## Kubernetes RBAC - denial of service risks

Reviewer: ```suggestion
## Kubernetes RBAC - denial of service risks {#denial-of-service-risks}
```

### Object creation denial-of-service {#object-creation-dos}
Users who have rights to create objects in a cluster may be able to create sufficient large 
objects to create a denial of service condition, as discussed in
[etcd used by Kubernetes is vulnerable to OOM attack](https://github.com/kubernetes/kubernetes/issues/107325). This may be
specifically relevant in multi-tenant clusters if semi-trusted or untrusted users 

> are allowed limited access to a system.

Reviewer 3: object size and object quantity are two dimensions that can trigger this. referencing resourcequota docs as a mitigation for object quantity could be useful

Author: updated this section .

Ideally minimal RBAC rights should be assigned to users and service accounts. Only permissions 
explicitly required for their operation should be used. Whilst each cluster will be different, 
some general rules that can be applied are :

 - Assign permissions at the namespace level where possible. Use RoleBindings as opposed to 
   ClusterRoleBindings to give users rights only within a specific namespace.
 - Avoid providing wildcard permissions when possible, especially to all resources.
   As Kubernetes is an extensible system, providing wildcard access gives rights
   not just to all object types presently in the cluster, but also to all future object types
   which are created in the future.
 - Administrators should not use `cluster-admin` accounts except where specifically needed. 
   Providing a low privileged account with [impersonation rights](/docs/reference/access-authn-authz/authentication/#user-impersonation)
   can avoid accidental modification of cluster resources.
 - Avoid adding users to the `system:masters` group. Any user who is a member of this group 
   bypasses all RBAC rights checks and will always have unrestricted superuser access, which cannot be 

>    revoked by removing Role Bindings or Cluster Role Bindings. As an aside, if a cluster is 

Reviewer: (nit)
```suggestion
   revoked by removing RoleBindings or ClusterRoleBindings. As an aside, if a cluster is
```

   using an authorization webhook, membership of this group also bypasses that webhook (requests 
   from users who are members of that group are never sent to the webhook)

### Minimize distribution of privileged tokens

> Ideally, pods shouldn't be assigned service accounts granted powerful permissions (listed [here](#Kubernetes-RBAC---Privilege-Escalation-Risks)). 

Reviewer: ```suggestion
Ideally, pods shouldn't be assigned service accounts that have been granted powerful permissions
(for example, any of the rights listed under
[privilege escalation risks](#privilege-escalation-risks)).
```
